### ISSUE TITLE
minification breaks library native is not defined.

### ISSUE THREAD

**Issue by IanSSenne** [PrismarineJS/prismarine-nbt#66]
I've build a small app using vite and when I run it in development it works fine but running it in production fails on load due to native not being defined. 
vite: https://vitejs.dev/
I have attached a small project that replicates this issue.
[vite-prismarine-nbt-bug1.zip](https://github.com/PrismarineJS/prismarine-nbt/files/8544716/prismarine-nbt-bug1.zip)

**Comment by IanSSenne** [comment]
one solution for this might be using new Function instead of eval and passing whatever is suppose to be native as a parameter.

**Comment by IanSSenne** [comment]
I was able to fix this by changing pnbt.js (generated using the command in the readme) by updating `Compiler.compile` to the following
```js
compile(code) {
                // Local variable to provide some context to eval()
                const native = this.native; // eslint-disable-line
                const { PartialReadError } = require("./utils"); // eslint-disable-line
                return new Function(
                  "native",
                  "PartialReadError",
                  "return (" + code + ")"
                )(native, PartialReadError)(); // eslint-disable-line
              }
              ```

**Comment by superx101** [comment]
I'm having the same problem with rollup.
I try solve it with the configuration:

```mjs
export default [
    {
        // ... other
        output: {
            preserveModules: true
        }
    }
]
```

or

```mjs
export default [
    {
        // ... other
        treeshake: false
    }
]
```

**Comment by LeoCaprile** [comment]
> I'm having the same problem with rollup. I try solve it with the configuration:
> 
> ```js
> export default [
>     {
>         // ... other
>         output: {
>             preserveModules: true
>         },
>         treeshake: false
>     }
> ]
> ```

this config solved your problem? I tried but with no success

**Comment by extremeheat** [comment]
node-protodef is compiling JS and running that with eval(), so the assumption is that the local variables are being captured by the eval(). Not fully sure what's going on here but it could be that perhaps the `native` var declaration is being deleted (as there are no references inside the codeblock) or it's being renamed somewhere.

Maybe doing something like a blank
```js
Object.getOwnPropertyDescriptors({ native })
```
statement could fix it

**Comment by LeoCaprile** [comment]
> node-protodef is compiling JS and running that with eval(), so the assumption is that the local variables are being captured by the eval(). Not fully sure what's going on here but it could be that perhaps the `native` var declaration is being deleted (as there are no references inside the codeblock) or it's being renamed somewhere.
> 
> Maybe doing something like a blank
> 
> ```js
> Object.getOwnPropertyDescriptors({ native })
> ```
> 
> statement could fix it

Can I open a PR adding this fix?

**Comment by superx101** [comment]
> > I'm having the same problem with rollup. I try solve it with the configuration:
> > ```js
> > export default [
> >     {
> >         // ... other
> >         output: {
> >             preserveModules: true
> >         },
> >         treeshake: false
> >     }
> > ]
> > ```
> 
> this config solved your problem? I tried but with no success

yes, this config did solved my problem

the problem I located looks like this:

```js
function foo(expression) {
     const native = this.native;
     eval(expression); // native is used here
}
```

but “native” is removed by rollup's default behavior

**Comment by extremeheat** [comment]
> Can I open a PR adding this fix?

You can open PR if you've tested it to work. The tree shaking functionality is in a way expected to cause problems because it's supposed to remove unused code. If this is the only place that there are problems I think it's ok to add a dummy statement (maybe even just `Object(native)` to handle this.

**Comment by IanSSenne** [comment]
> > Can I open a PR adding this fix?
> 
> You can open PR if you've tested it to work. The tree shaking functionality is in a way expected to cause problems because it's supposed to remove unused code. If this is the only place that there are problems I think it's ok to add a dummy statement (maybe even just `Object(native)` to handle this.

I don't think this is sufficient due to minification being able to change the variable name, I think the best option is to wrap the code into a function instead of eval so the values can be named as arguments, the downside of this is some work will have to be done to make sure the code generated does not depend on things other than native as they will have to be specifically included in that scope
